src/libraries/System.Security.Cryptography/tests/OpenSslNamedKeysTests.manual.cs

```diff
@@ -174,17 +225,202 @@ public static void Engine_UsePublicKey()
             using ECDsa ecdsaPri = new ECDsaOpenSsl(priKeyHandle);
             using ECDsa ecdsaBad = ECDsa.Create();
             ecdsaBad.KeySize = ecdsaPri.KeySize;
 
             byte[] data = new byte[] { 1, 2, 3, 1, 1, 2, 3 };
             byte[] signature = ecdsaPri.SignData(data, HashAlgorithmName.SHA256);
             byte[] badSignature = ecdsaBad.SignData(data, HashAlgorithmName.SHA256);
+            Assert.Equal(signature.Length, badSignature.Length);
             Assert.NotEqual(data, signature);
             Assert.True(ecdsaPri.VerifyData(data, signature, HashAlgorithmName.SHA256));
             Assert.False(ecdsaPri.VerifyData(data, badSignature, HashAlgorithmName.SHA256));
         }
+
+        [ConditionalFact(nameof(ShouldRunProviderEcDsaTests))]
+        public static void Provider_TPM2ECDSA()
+        {
+            using SafeEvpPKeyHandle priKeyHandle = SafeEvpPKeyHandle.OpenKeyFromProvider(Tpm2ProviderName, TpmEcDsaKeyHandleUri);
+            using ECDsa ecdsaPri = new ECDsaOpenSsl(priKeyHandle);
+
+            byte[] data = new byte[] { 1, 2, 3, 1, 1, 2, 3 };
+
+            byte[] ecdsaPubBytes = ecdsaPri.ExportSubjectPublicKeyInfo();
+            ECDsa ecdsaPub = ECDsa.Create();
+            ecdsaPub.ImportSubjectPublicKeyInfo(ecdsaPubBytes, out int bytesRead);
+            Assert.Equal(ecdsaPubBytes.Length, bytesRead);
+
+            using ECDsa ecdsaBad = ECDsa.Create();
+            ecdsaBad.KeySize = ecdsaPri.KeySize;
+
+            // Verify can sign/verify multiple times
+            for (int i = 0; i < 10; i++)
+            {
+                data[0] = (byte)i;
+                byte[] signature = ecdsaPri.SignData(data, HashAlgorithmName.SHA256);
+                byte[] badSignature = ecdsaBad.SignData(data, HashAlgorithmName.SHA256);
+                Assert.NotEqual(data, signature);
+                Assert.NotEqual(data, badSignature);
+                Assert.NotEqual(badSignature, signature);
+                Assert.True(ecdsaPub.VerifyData(data, signature, HashAlgorithmName.SHA256));
+                Assert.False(ecdsaPub.VerifyData(data, badSignature, HashAlgorithmName.SHA256));
+                Assert.False(ecdsaBad.VerifyData(data, signature, HashAlgorithmName.SHA256));
+
+                // TPM key is intended for sign/decrypt only, we could theoretically make verify work without needing to export/import by forcing 'default' provider
+                // for this operation but it's most likely misusage on user part and tpm2 provider intentionally didn't allow it so we will follow this logic.
+                Assert.ThrowsAny<CryptographicException>(() => ecdsaPri.VerifyData(data, signature, HashAlgorithmName.SHA256));
+            }
+
+            // It's TPM so it should not be possible to export parameters
+            Assert.ThrowsAny<CryptographicException>(() => ecdsaPri.ExportParameters(includePrivateParameters: true));
```

Comment:
Probably good to assert that it can export public.  And then perhaps explicit public (and if that doesn't work... make it work :smile:)

Reply:
added separate tests for that